clang/compiler-course/kudryashova_i_compilers_clangast/kudryashova_i_clangast_plugin.cpp

```diff
@@ -0,0 +1,104 @@
+#include "clang/AST/ParentMapContext.h"
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+
+class ImplicitConvVisitor 
+  : public clang::RecursiveASTVisitor<ImplicitConvVisitor> {
+	  
+private:
+  clang::ASTContext *m_context;
+  std::map<const clang::FunctionDecl*, 
+           std::map<std::pair<std::string, std::string>, int>> m_functionStats;
+  int m_totalConversions = 0;
+  
+public:
+  explicit ImplicitConvVisitor(clang::ASTContext *context) 
+    : m_context(context) {}
+
+  bool VisitImplicitCastExpr(clang::ImplicitCastExpr *ICE) {
+	auto castKind = ICE->getCastKind();
+    if (castKind == clang::CK_LValueToRValue || 
+      castKind == clang::CK_NoOp ||
+      castKind == clang::CK_FunctionToPointerDecay) {
+      return true;
+    }
+    clang::QualType SourceType = ICE->getSubExpr()->getType();
+    clang::QualType TargetType = ICE->getType();
+	
+	if (SourceType.getAsString() == TargetType.getAsString()) {
+      return true;
+    }
+
+    auto Node = clang::DynTypedNode::create<clang::ImplicitCastExpr>(*ICE);
+    auto Parents = m_context->getParents(Node);
+    
+    while (!Parents.empty()) {
+      if (const clang::FunctionDecl *FD = 
+          Parents[0].get<clang::FunctionDecl>()) {
+          std::string From = SourceType.getAsString();
+          std::string To = TargetType.getAsString();
+            
+          m_functionStats[FD][std::make_pair(From, To)]++;
+          m_totalConversions++;
+          break;
+        }
+
+      Node = Parents[0];
+      Parents = m_context->getParents(Node);
+    }
+    return true;
+  }
+
+
+  void printStats(llvm::raw_ostream &OS) {
+    for (const auto &funcEntry : m_functionStats) {
+      OS << "Function `" << funcEntry.first->getName() << "`\n";
+      for (const auto &conv : funcEntry.second) {
+        OS << conv.first.first << " -> " 
+           << conv.first.second << ": " 
+           << conv.second << "\n";
+      }
+    }
```

Comment:
I think it will be more readable
```suggestion
    for (const auto &[func, convs] : m_functionStats) {
      OS << "Function `" << func->getName() << "`\n";
      for (const auto &[conv, num] : convs) {
        OS << conv.first << " -> " 
           << conv.second << ": " 
           << num << "\n";
      }
    }
```